docs/devices.ipynb

```diff
@@ -0,0 +1,357 @@
+    "except Exception as e:\n",
+    "  print(e)\n",
+    "\n",
+    "try:\n",
+    "  my_device.validate_operation(op2)\n",
+    "except Exception as e:\n",
+    "  print(e)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "EcTgQg5xCxzG"
+   },
+   "source": [
+    "These validation operations can also be used with full circuits:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "EiolyMKLC262"
+   },
+   "outputs": [],
+   "source": [
+    "my_circuit = cirq.Circuit(\n",
+    "    cirq.PhasedXPowGate(phase_exponent=0.3)(q1),\n",
+    "    cirq.PhasedXPowGate(phase_exponent=0.3)(q2),\n",
+    "    cirq_google.SYC(q1, q2),\n",
+    "    cirq_google.SYC(q2, q3),\n",
+    ")\n",
+    "my_device.validate_circuit(my_circuit)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "-sNjRTCKDb2D"
+   },
+   "source": [
+    "`my_circuit` satisfies all the device constraints and could be run on a Sycamore device."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "BrxB0eCUDvlI"
+   },
+   "source": [
+    "## 2. Metadata features\n",
+    "\n",
+    "Some devices will also expose additional information via the `metadata` property. You can look at the metadata information of the Sycamore device with:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "clsvdcXzEOIV"
+   },
+   "outputs": [],
+   "source": [
+    "metadata = my_device.metadata\n",
+    "metadata"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "glMmeLevElt-"
+   },
+   "source": [
+    "The Sycamore device is a 2d grid device that exposes a `cirq.GridDeviceMetadata` with a uniform set of gates across all the qubits as well as a connectivity graph. You can explore the properties below:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "9Zm_1IsFEiHT"
+   },
+   "outputs": [],
+   "source": [
+    "type(metadata)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "JbwrPD_AETiX"
+   },
+   "outputs": [],
+   "source": [
+    "qubit_set = metadata.qubit_set\n",
+    "nx_graph = metadata.nx_graph\n",
+    "gateset = metadata.gateset\n",
+    "\n",
+    "print(gateset)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "fo6ZTYN4HBVL"
+   },
+   "source": [
+    "These metadata features can be useful when designing/building algorithms around certain device information in order to tailor them for that device."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "RvC4xz_OHT3s"
+   },
+   "source": [
+    "## 3. The `cirq.Device` interface\n",
+    "\n",
+    "For advanced users (such as vendors) it is also possible to implement your own Device with its own unique constraints and metadata information. Below you can implement our own fictituous device:"
```

Comment:
It might be good to put some more information here about what to do in order to implement your own device.
For instance, "The minimum you will need to do is write an __init__ to initialize the object and a validate_operation function to validate a single operation.   However, this interface supports constraints like not allowing adjacent operations with validate_moment, or even circuit level constrains like a maximum depth using validate_circuit.  You can also add metadata to your object by subclassing DeviceMetadata."  (etc)

Reply:
I agree with this. Details of the function types of `init` and `validate_operation`, as well as the types of data that can be stored in `DeviceMetadata` would be great to have in this doc, so people don't have to go to the reference if they're trying to match this template.

Reply:
Expanded the wording as also added metadata example for the custom case.

Reply:
The docstrings for the below Device are looking great.  I think I would still expand this paragraph to give some motivation as to the following class.  For instance, 
"Suppose we have a 5 qubit device that supports X/Y/Z single qubit and CZ gates between neighbors with a maximum of 10 CZ gates.  We would then create an instance of the Device class that has a validate_operation to validate the gate types and adjacency requirements, and implement a validate_circuit to validate the maximum number of CZs."

Reply:
Added this sentence in.